### ISSUE TITLE
When data is too big

### ISSUE THREAD

**Issue by georgemilosh** [georgemilosh/Climate-Learning#50]
There are not so many machines that can handle all the data in memory. We need to find a way to deal with this issue. There are many materials online which show how it can be achieved in tensorflow (pipelines etc). The problem is that I still get `out of memory` issue (the process simply gets killed).

**Comment by georgemilosh** [comment]
I should note that there are extra memory leaks when trying current tensorflow version (that is no longer tensorflow-gpu and that is installed with pip rather than conda)

**Comment by AlessandroLovo** [comment]
Check for easy improvement when loading data, especially if we don't want to use the whole dataset

**Comment by georgemilosh** [comment]
At the moment I create a separate dataset with `cdo` that's the subset, and everything works much faster. But I feel this is ad hoc. Alternatively we could always work with smaller datasets that are to be concatenated in Learn2_new.py if needed, but also a bit annoying

**Comment by AlessandroLovo** [comment]
Yeah, indeed. I was planning to do line by line evaluation of the code monitoring the RAM to see exactly when we load the data into memory and if we can do something about it, but I cannot guarantee I'll have time for that

**Comment by AlessandroLovo** [comment]
Found the reason: before the network had ~200.000 trainable parameters... now it has ~8 millions...

**Comment by AlessandroLovo** [comment]
And that is because the MaxPool layers have disappeared

**Comment by georgemilosh** [comment]
Although virtual memory normally doesn't matter.

**Comment by AlessandroLovo** [comment]
I will re-implement the MaxPool for backward compatibility

**Comment by georgemilosh** [comment]
My view is that we will need to implement tensorflow datasets. Operations such as shuffling (balancing) and train/validation have to be done virtually by permuting only indices. The dataset has to somehow know which portions of the full dataset it has to provide for the `next(batch)`. In pytorch it is quite easy to control how data is extracted. I don't like that we have to copy paste data in memory. Normalization is another step, but it could be achieved by a layer implementation rather than doing it explicitly like we do?
